src/devices/machine/t10mmc.cpp

```diff
@@ -506,6 +506,49 @@ void t10mmc::ExecCommand()
 		m_transfer_length = 0;
 		break;
 
+	case T10MMC_CMD_READ_CD: {
+		// TODO: Implement reladr bit, flag bits, test and handle other conditions besides reads to "any type" sector types
+		m_lba = command[2]<<24 | command[3]<<16 | command[4]<<8 | command[5];
+		m_blocks = command[6]<<16 | command[7]<<8 | command[8];
+
+		// m_device->logerror("T10MMC: READ CD start_lba[%08x] block_len[%06x] %02x %02x %02x %02x\n", m_lba, m_blocks, command[1], command[9], command[10], command[11]);
+
+		auto expected_sector_type = BIT(command[1], 2, 3);
+		if (expected_sector_type != 0) {
+			m_device->logerror("T10MMC: READ CD requested a sector type of %d which is unhandled\n", expected_sector_type);
+
+			auto trk = m_cdrom->get_track(m_lba);
+			auto track_type = m_cdrom->get_track_type(trk);
+			if ((expected_sector_type == 1 && track_type != cdrom_file::CD_TRACK_AUDIO)
+			|| (expected_sector_type == 2 && track_type != cdrom_file::CD_TRACK_MODE1 && track_type != cdrom_file::CD_TRACK_MODE1_RAW)
+			|| (expected_sector_type == 3 && track_type != cdrom_file::CD_TRACK_MODE2 && track_type != cdrom_file::CD_TRACK_MODE2_RAW)
+			|| (expected_sector_type == 4 && track_type != cdrom_file::CD_TRACK_MODE2_FORM1)
+			|| (expected_sector_type == 5 && track_type != cdrom_file::CD_TRACK_MODE2_FORM2)) {
+				set_sense(SCSI_SENSE_KEY_ILLEGAL_REQUEST, SCSI_SENSE_ASC_ASCQ_ILLEGAL_MODE_FOR_THIS_TRACK);
+
+				m_phase = SCSI_PHASE_STATUS;
+				m_status_code = SCSI_STATUS_CODE_CHECK_CONDITION;
+				m_transfer_length = 0;
+				break;
+			}
+		}
```

Comment:
Not sure I follow the logic here, gonna checkout documentation later:

1. What happens if the condition gets true, is it intended behaviour that is gonna move on and fail miserably at the execute command step below, hard crash at worst?
2. The if block itself looks a good place to functional programming, as it stands is pretty unreadable.

Reply:
I guess I can remove this whole block and just give an error for any mode that hasn't been tested yet. The check itself is this. The expected sector type and the type of the track being read must match or else the error condition should be returned as shown in the code. The check itself really does boil down to what you see above. The rest should occur in the read data function where it should properly handle reading data for that track type and return it in the appropriate format.

<img width="1056" alt="image" src="https://user-images.githubusercontent.com/63495610/233411322-290b51f4-f40f-4256-9b74-143f0bb04eea.png">

Reply:
I could read that particular piece of code fine, and it agrees with what the command should do (i.e. return an error if the read mode is inappropriate for the track mode).

Reply:
I added in stricter checks for things that haven't been tested yet. Non-mode 1 tracks for example would probably fail, but I don't know of anything else that uses this command to be able to test it more.

Reply:
I see, so it's supposed to check for the sector type _then_ propagate during command execution, because it can potentially do track trespassing that have diverging modes. I guess I'll get there once I have better grasp of reading sector modes.